lib/srv/db/elasticsearch/engine.go

```diff
@@ -171,10 +186,170 @@ func (e *Engine) process(ctx context.Context, sessionCtx *common.Session, req *h
+		"_knn_search",    // https://www.elastic.co/guide/en/elasticsearch/reference/master/knn-search-api.html
+		"_msearch",       // https://www.elastic.co/guide/en/elasticsearch/reference/master/multi-search-template.html, https://www.elastic.co/guide/en/elasticsearch/reference/master/search-multi-search.html
+		"_search_shards", // https://www.elastic.co/guide/en/elasticsearch/reference/master/search-shards.html
+		"_count",         // https://www.elastic.co/guide/en/elasticsearch/reference/master/search-count.html
+		"_validate",      // https://www.elastic.co/guide/en/elasticsearch/reference/master/search-validate.html
+		"_terms_enum",    // https://www.elastic.co/guide/en/elasticsearch/reference/master/search-terms-enum.html
+		"_explain",       // https://www.elastic.co/guide/en/elasticsearch/reference/master/search-explain.html
+		"_field_caps",    // https://www.elastic.co/guide/en/elasticsearch/reference/master/search-field-caps.html
+		"_rank_eval",     // https://www.elastic.co/guide/en/elasticsearch/reference/master/search-rank-eval.html
+		"_mvt":           // https://www.elastic.co/guide/en/elasticsearch/reference/master/search-vector-tile-api.html
+		return parts[1], apievents.ElasticsearchCategory_SEARCH
+	}
+
+	return "", apievents.ElasticsearchCategory_GENERAL
+}
+
+// getQueryFromRequestBody attempts to find the actual query from the request body, to be shown to the interested user.
+func (e *Engine) getQueryFromRequestBody(contentType string, body []byte) string {
+	var q struct {
+		Query any `json:"query" yaml:"query"`
+		Knn   any `json:"knn" yaml:"knn"`
+	}
+
+	switch contentType {
+	// CBOR and Smile are officially supported by Elasticsearch:
+	// https://www.elastic.co/guide/en/elasticsearch/reference/master/api-conventions.html#_content_type_requirements
+	// We don't support introspection of these content types, at least for now.
+	case "application/cbor":
+		e.Log.Warnf("Content type not supported: %q.", contentType)
+		return ""
+
+	case "application/smile":
+		e.Log.Warnf("Content type not supported: %q.", contentType)
+		return ""
+
+	case "application/yaml":
+		err := yaml.Unmarshal(body, &q)
+		if err != nil {
+			e.Log.WithError(err).Warnf("Error decoding request body as %q.", contentType)
+			return ""
+		}
+
+	case "application/json":
+		err := json.Unmarshal(body, &q)
+		if err != nil {
+			e.Log.WithError(err).Warnf("Error decoding request body as %q.", contentType)
+			return ""
+		}
+
+	default:
+		e.Log.Warnf("Unknown or missing 'Content-Type': %q, assuming 'application/json'.", contentType)
+		err := json.Unmarshal(body, &q)
+		if err != nil {
+			e.Log.WithError(err).Warnf("Error decoding request body as %q.", contentType)
+			return ""
+		}
+	}
+
+	result := q.Query
+	if result == nil {
+		result = q.Knn
+	}
+
+	if result == nil {
+		return ""
+	}
+
+	switch qt := result.(type) {
+	case string:
+		return qt
+	default:
+		marshal, err := json.MarshalIndent(result, "", "   ")
```

Comment:
Do we need "Indent"? how does it look on the web app?

Reply:
We probably don't and shoudln't: barely helps for most users and is more verbose.